### PR TITLE
Ensure that, when a dialog is reopened, only columns that still exist are shown in the list of selected columns

### LINKED ISSUE #7577: Ensure multiple receiver clears on reopen if the column no longer exists
We now have a partial fix - the single receiver clears on reopen if the column no longer exists.

Still to do:
- Checks columns exists on reopening in multiple receiver
- Check the column satisfies metadata conditions for single & multiple receivers on reopen
- Implement checks when receiver is for other types - not only columns

_Originally posted by @dannyparsons in https://github.com/africanmathsinitiative/R-Instat/issues/1504#issuecomment-312699964_

### PR DESCRIPTION
**Pull request by N-thony**
Fixes #7577 
@africanmathsinitiative/developers @rdstern @Amsale2022 this is ready for review. Thanks.

### CODE REVIEW

**Comment by Patowhiz** [comment]
@N-thony could you explore getting this information from the .Net databook.
One of the objectives in refactoring current R-Instat implementations is to have very minimal interactions with R especially when loading dialogs. It helps with making R-Instat work optimally with both wide and long datasets.

I understand there may be no structures in place yet for the above functionality. If so and it needs more discussions, I'm happy if a to do comment linking to this PR or issue is added. So that in future once those structures are added we can easily refactor the code added here.
In addition, I have a similar draft PR #7440 that attempts to add some of these structures. It's far from being ready but demonstrates my line of thought.

On a separate note, kindly use a wide data set when testing this addition. To just make sure we don't introduce regression.

**Review by lloyddewit — DISMISSED** [review]

**Comment by N-thony** [comment]
> @N-thony Thank you for the changes, I approved. Regarding @Patowhiz 's comments, do you plan to address these in this PR (i.e. by deleting the column using VB rather than R), or do you prefer to do this in future in a separate PR? If you prefer a separate PR, then please could you raise an issue for this that references this PR (#7581)?
> 
> @rdstern Please could you test?
> 
> Thanks!

@lloyddewit I have opened an issue for the improvement of this, also I did test with a large dataset and it seems to work fine.

**Comment by rdstern** [comment]
@Patowhiz are you now ok with this one?

**Comment by rdstern** [comment]
@Patowhiz are you now ok with this change?

**Comment by Patowhiz** [comment]
@N-thony @lloyddewit @rdstern below is my feedback.

I have tested this PR using poulet dataset in `factoMineR` package. It has 43 rows and 7407 variables. We have it in our library.

The code additions done here will result to a performance regression when it comes to reloading dialogs when working with such similar wide data sets.

The `Add` subroutine, changed here, is called for every variable read from the R script function during `SetRCode`. So if we have 7407 variables in the R script function, the code added here goes back to R, 7407 times. We definitely don't want that.
This is what the selector used to do before we refactored it. 

Also note, the selector already clears the column that doesn't not exist, I wonder if you could explore using it to clear the columns in this receiver. The 2 are already linked anyway, so we just need to improve that linking in away that variable in the receiver are consistent with variable in the selector. 

I understand that route is also not so clear because the current selector and receiver link needs to be refactored too!

**Comment by rdstern** [comment]
@N-thony given the comments above, what is to be done with this.  Do we just close this pull request for now?

**Comment by Patowhiz** [comment]
@rdstern @N-thony 
I would suggest to do this in the following steps;

1. Find a way for the specific metadata changes at R level to be detected at .Net level. this could be implemented in a general way, to cover specifc metadata change.
2. Improve the .Net data book metadata implementation. It should contain enough metadata information about the data frame in a way that the controls read from it. 
3. Make the selector and receivers read data from the .Net data book. Refreshing of selectors and receivers will no longer have a huge performance overhead. This will delink the selectors and receivers from R. Which is want we are consistently trying to do!
4. Finally solve the issues outlined here.

@rdstern if we solve the above issues, we can have dialogs be "overlays".

**Comment by rdstern** [comment]
@N-thony I wonder what should happen here.  I note, from @Patowhiz this links to the topic of wide data sets, and I hope there can be work on their efficient handling soon.  It is now almost a year since Chris and Patrick were of the view that we should be able to cope with data sets of almost any width?

**Comment by Patowhiz** [comment]
@rdstern as I [commented](https://github.com/africanmathsinitiative/R-Instat/pull/7581#issuecomment-1379267994) above, to do this correctly, we have to refactor the metadata change detection at both .Net and R level. And the receivers should read the  metadata from the .Net data book structures.

**Comment by lloyddewit** [comment]
@rdstern @Patowhiz @N-thony 

I understand from the comments above, that the solution in this PR will not work well for wide data sets. 

@Patowhiz stated:

> The Add subroutine, changed here, is called for every variable read from the R script function during SetRCode. So if we have 7407 variables in the R script function, the code added here goes back to R, 7407 times. We definitely don't want that. 

On 24 Nov, @Patowhiz suggested doing the work at the dotnet level (see above). This would require deeper architectural understanding and is dependent on other refactorings. We may have to wait a long time for this to be implemented.

As a compromise, could we use the R function that @N-thony developed but ensure that we only call it once (rather than 7407 times)? Could we achieve this by moving the R code out of the `Add()` sub and then calling it just once for the entire list of column names (i.e. set the `col_names` parameter to be a list of columns)? If this approach won't work, is there another approach using R that would work with a single R function call?

If doing this in dotnet is the only fast enough solution, then we may need to close this PR and reassign issue #7594 and issue #7577 .

What are your views?

**Comment by N-thony** [comment]
> 

@lloyddewit thanks for this, I will try your approach first and let you know how it goes.

**Comment by N-thony** [comment]
@lloyddewit as discussed I found that we don't need to use R since the selector list will always contains the updated list of variable. As the code need to be refactor as suggested above, this seems to be a temporary solution. Please have a look.
@rdstern you can also test if everything is okay?

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@N-thony Thank you for coming up with this idea. I find it a simple and pragmatic solution.
Just a few small comments.

File: instat/ucrReceiverMultiple.vb [hunk 1]
```diff
@@ -309,6 +308,18 @@ Public Class ucrReceiverMultiple
         lstSelectedVariables.BackColor = Color.White
     End Sub
 
+    Public Overrides Sub RemoveVariablesInReceiver(lstViewItem As ListView)
```
[the remark this answers is not in the document]
Comment (N-thony) [reply instat/ucrReceiverMultiple.vb:311]: I think the name you have proposed is clear, I like it. I added the XML header too.

File: instat/ucrReceiver.vb [hunk 2]
```diff
@@ -130,6 +130,11 @@ Public Class ucrReceiver
 
     End Sub
 
+    Public Overridable Sub RemoveVariablesInReceiver(lstViewItem As ListView)
```
[the remark this answers is not in the document]
Comment (N-thony) [reply instat/ucrReceiver.vb:133]: yes, if I delete this sub, we do get the build errors.

**Comment by N-thony** [comment]
@lloyddewit please take a look, I resolved your comments above.

**Review by lloyddewit — APPROVED** [review]
@N-thony excellent, thanks

**Comment by lloyddewit** [comment]
@rdstern there are some code changes since your last approval. If you can retest/approve, then we can merge.
Thanks

**Review by rdstern — APPROVED** [review]
Nice

File: instat/ucrReceiverMultiple.vb [hunk 3]
```diff
@@ -309,6 +308,19 @@ Public Class ucrReceiverMultiple
         lstSelectedVariables.BackColor = Color.White
     End Sub
 
+    ''' <summary>
+    ''' Removes any variable in the multiple receiver
+    ''' that is not in the list of variables of the selector
+    ''' </summary>
+    ''' <param name="lstViewItem"></param>
+    Public Overrides Sub RemoveAnyVariablesNotInList(lstViewItem As ListView)
```
Comment (Patowhiz) [inline instat/ucrReceiverMultiple.vb:316]: @N-thony @lloyddewit I was expecting to see this implementation done for the single receivers as well. Or did they not need this implementation. The block at the `ucrReceiver` is empty.

File: instat/ucrReceiver.vb [hunk 4]
```diff
@@ -130,6 +130,10 @@ Public Class ucrReceiver
 
     End Sub
 
+    Public Overridable Sub RemoveAnyVariablesNotInList(lstViewItem As ListView)
```
Comment (Patowhiz) [inline instat/ucrReceiver.vb:133]: @N-thony @lloyddewit It also seems a bit odd to have the list view as a parameter of the `RemoveAnyVariablesNotInList()`, considering all receivers already have link to the selector. We could just use the selector variable define in line 22. In addition, we wouldn't want this function be used to remove variables from any other list view other than the linked selector.

**Review by Patowhiz — COMMENTED** [review]
@N-thony @lloyddewit many thanks for this.
As a temporary implementation, I agree this will suffice.
I only have a few comments regarding the implementation. See my review.

In addition, @N-thony I think it would have been good if you could add a few comments that mention this PR and it's comments. I tend to think, it's a good to declare all temporary solutions in the comments. 
As @rdstern has mentioned before, it is now becoming critical that we correctly implement working with wide data sets.

Thanks.

**Comment by lloyddewit** [comment]
@Patowhiz Thank you for the review, and helpful comments.

> As a temporary implementation, I agree this will suffice.

This solution checks that the variables listed in the receiver currently exist in the selector. I agree that this assumes that the selector lists the correct variables but the receiver already depends on the selector so I think this is a reasonable assumption. For me this PR is a simple and pragmatic solution. Hopefully the receivers will be refactored at one point, but I don't see the solution in this PR as more 'temporary' than other code.

@N-thony The comments above look like good code quality improvements. If you agree they are worth adding, then will you do this in a separate PR?
Thanks for your help and patience

File: instat/ucrSelector.vb [hunk 5]
```diff
@@ -148,6 +148,7 @@ Public Class ucrSelector
 
         'todo, for columns, the list view should be field with variables from the .Net metadata object
         frmMain.clsRLink.FillListView(lstAvailableVariable, strType:=strCurrentType, lstIncludedDataTypes:=lstCombinedMetadataLists(0), lstExcludedDataTypes:=lstCombinedMetadataLists(1), strHeading:=CurrentReceiver.strSelectorHeading, strDataFrameName:=strCurrentDataFrame, strExcludedItems:=arrStrExclud, strDatabaseQuery:=CurrentReceiver.strDatabaseQuery, strNcFilePath:=CurrentReceiver.strNcFilePath)
+        CurrentReceiver.RemoveAnyVariablesNotInList(lstAvailableVariable)
         EnableDataOptions(strCurrentType)
```
[the remark this answers is not in the document]
Comment (Patowhiz) [reply instat/ucrSelector.vb:152]: In addition, should this happen for all receivers or just the current receiver? If all, then I think the linked receivers are the ones that should listen to selector variables changes and remove the variables they contain but are not in the selector.

**Comment by Patowhiz** [comment]
@lloyddewit you are correct, it's not more 'temporary' than other code.
Thanks.

**Pull request merged**